### ISSUE TITLE
Breaking the guarantees of 'Path Rel Dir'

### ISSUE THREAD

**Issue by hasufell** [commercialhaskell/path#18]
In ghci:

``` hs
Prelude Path> dir <- parseAbsDir  "/"
Prelude Path> dirname dir
"/"
Prelude Path> :t dirname dir
dirname dir :: Path Rel Dir
```

That means I have a relative dir that is called `/`.

But when I try to parse it, it fails:

``` hs
Prelude Path> parseRelDir "/"
*** Exception: InvalidRelDir "/"
```

So I've just constructed a `Path Rel Dir` via the library functions that is not supposed to be a relative dir.

**Comment by hasufell** [comment]
I think the only reliable way to fix that is to break API and do something along these lines:
```hs
dirname :: MonadThrow m => Path b Dir -> m (Path Rel Dir)
dirname (Path l)
  | not (FilePath.isAbsolute rl) = return $ Path rl
  | otherwise                    = throwM (InvalidRelDir rl)
  where rl = last (FilePath.splitPath l)
```

**Comment by chrisdone** [comment]
Good point. Maybe `dirname` should be split into `relDirname` to get the dirname of a relative directory which is always valid, and `absDirName` which is not guaranteed.

**Comment by hasufell** [comment]
Another way could be to add another Dir type that expresses the directory is a non-root directory. The `/` dir is always a bit special, but I'm not sure if that's a good reason to encode that information on type level. It'll also be quite intrusive to the API.

**Comment by chrisdone** [comment]
I think `dirname`, like `parent` (#33), being a breaking change sounds like a reasonable and less intrusive change. 

I propose we plan a migration path for changing this function's type.

**Comment by mrkkrp** [comment]
Let's put the change on 0.6.0 milestone? As I understand the consensus is that `dirname` should live in `MonadThrow`?

**Comment by NorfairKing** [comment]
>  I understand the consensus is that dirname should live in MonadThrow?
+1

**Comment by mrkkrp** [comment]
Hopefully fixed by #34.

**Comment by hasufell** [comment]
Neat trick. Although `/` technically (by POSIX standard and filesystem technicalities) really has no _name_.
